File: TimberCruisingApp/InventoryEngine/VolumeEquations/TableLookup.swift
// Spec §7.7. User-provided volume table with bilinear interpolation on
// (DBH, H). The coefficient record encodes the grid as a flat dictionary:
//
//   "dbh_0", "dbh_1", ..., "dbh_{n-1}"         monotonic DBH grid (cm)
//   "h_0",   "h_1",   ..., "h_{m-1}"           monotonic H grid (m)
//   "v_{i}_{j}" for i in 0..<n, j in 0..<m     V(cm, m) = m³
//
// Out-of-range queries clamp to the nearest edge.

import Foundation

public struct TableLookup: VolumeEquation {

    private let dbhGrid: [Float]     // ascending cm
    private let hGrid:   [Float]     // ascending m
    private let volumes: [[Float]]   // [i][j] = V at (dbhGrid[i], hGrid[j])
    private let merchFraction: Float

    public init(coefficients: [String: Float]) {
        let dbhCount = Self.count(prefix: "dbh_", in: coefficients)
        let hCount   = Self.count(prefix: "h_",   in: coefficients)
        guard dbhCount >= 2, hCount >= 2 else {
            fatalError("TableLookup needs at least 2 DBH and 2 H grid points")
        }
        let dbh = (0..<dbhCount).map { CoefficientLookup.required(coefficients, "dbh_\($0)") }
        let h   = (0..<hCount).map   { CoefficientLookup.required(coefficients, "h_\($0)") }
        var vols: [[Float]] = Array(repeating: Array(repeating: 0, count: hCount),
                                    count: dbhCount)
        for i in 0..<dbhCount {
            for j in 0..<hCount {
                vols[i][j] = CoefficientLookup.required(coefficients, "v_\(i)_\(j)")
            }
        }
        self.dbhGrid = dbh
        self.hGrid = h
        self.volumes = vols
        self.merchFraction = CoefficientLookup.optional(coefficients, "merchFraction", default: 0.85)
    }

    public func totalVolumeM3(dbhCm: Float, heightM: Float) -> Float {
        guard dbhCm > 0, heightM > 0 else { return 0 }
        let (i0, i1, tI) = Self.bracket(dbhCm, in: dbhGrid)
        let (j0, j1, tJ) = Self.bracket(heightM, in: hGrid)
        let v00 = volumes[i0][j0]
        let v01 = volumes[i0][j1]
        let v10 = volumes[i1][j0]
        let v11 = volumes[i1][j1]
        let v0 = v00 + (v01 - v00) * tJ
        let v1 = v10 + (v11 - v10) * tJ
        return v0 + (v1 - v0) * tI
    }

    public func merchantableVolumeM3(dbhCm: Float, heightM: Float,
                                     topDibCm: Float, stumpHeightCm: Float) -> Float {
        totalVolumeM3(dbhCm: dbhCm, heightM: heightM) * merchFraction
    }

    // MARK: - helpers

    /// Number of sequential keys `<prefix>0, <prefix>1, ...` present in `dict`.
    private static func count(prefix: String, in dict: [String: Float]) -> Int {
        var n = 0
        while dict["\(prefix)\(n)"] != nil { n += 1 }
        return n
    }

    /// Return (low-index, high-index, t) for bracketing `x` within `grid`.
    /// Clamps at both ends.
    private static func bracket(_ x: Float, in grid: [Float]) -> (Int, Int, Float) {
        if x <= grid.first! { return (0, min(1, grid.count - 1), 0) }
        if x >= grid.last!  { let last = grid.count - 1
                              return (max(0, last - 1), last, 1) }
        var lo = 0, hi = grid.count - 1
        while hi - lo > 1 {
            let mid = (hi + lo) / 2
            if grid[mid] <= x { lo = mid } else { hi = mid }
        }
        let t = (x - grid[lo]) / (grid[hi] - grid[lo])
        return (lo, hi, t)
    }
}

File: android/app/src/main/java/com/hcjeong/forestix/inventory/TableLookup.kt
// Port of iOS InventoryEngine/VolumeEquations/TableLookup.swift.
// Spec §7.7. User-provided volume table with bilinear interpolation on
// (DBH, H). The coefficient record encodes the grid as a flat dictionary:
//
//   "dbh_0", "dbh_1", ..., "dbh_{n-1}"         monotonic DBH grid (cm)
//   "h_0",   "h_1",   ..., "h_{m-1}"           monotonic H grid (m)
//   "v_{i}_{j}" for i in 0..<n, j in 0..<m     V(cm, m) = m³
//
// Out-of-range queries clamp to the nearest edge.

package com.hcjeong.forestix.inventory

import kotlin.math.max
import kotlin.math.min

class TableLookup(coefficients: Map<String, Float>) : VolumeEquation {

    private val dbhGrid: List<Float>     // ascending cm
    private val hGrid: List<Float>       // ascending m
    private val volumes: Array<FloatArray>   // [i][j] = V at (dbhGrid[i], hGrid[j])
    private val merchFraction: Float

    init {
        val dbhCount = count(prefix = "dbh_", dict = coefficients)
        val hCount = count(prefix = "h_", dict = coefficients)
        require(dbhCount >= 2 && hCount >= 2) {
            "TableLookup needs at least 2 DBH and 2 H grid points"
        }
        dbhGrid = (0 until dbhCount).map { CoefficientLookup.required(coefficients, "dbh_$it") }
        hGrid = (0 until hCount).map { CoefficientLookup.required(coefficients, "h_$it") }
        volumes = Array(dbhCount) { i ->
            FloatArray(hCount) { j ->
                CoefficientLookup.required(coefficients, "v_${i}_$j")
            }
        }
        merchFraction = CoefficientLookup.optional(coefficients, "merchFraction", default = 0.85f)
    }

    override fun totalVolumeM3(dbhCm: Float, heightM: Float): Float {
        if (dbhCm <= 0f || heightM <= 0f) return 0f
        val (i0, i1, tI) = bracket(dbhCm, grid = dbhGrid)
        val (j0, j1, tJ) = bracket(heightM, grid = hGrid)
        val v00 = volumes[i0][j0]
        val v01 = volumes[i0][j1]
        val v10 = volumes[i1][j0]
        val v11 = volumes[i1][j1]
        val v0 = v00 + (v01 - v00) * tJ
        val v1 = v10 + (v11 - v10) * tJ
        return v0 + (v1 - v0) * tI
    }

    override fun merchantableVolumeM3(dbhCm: Float, heightM: Float,
                                      topDibCm: Float, stumpHeightCm: Float): Float =
        totalVolumeM3(dbhCm = dbhCm, heightM = heightM) * merchFraction

    // MARK: - helpers

    private companion object {

        /// Number of sequential keys `<prefix>0, <prefix>1, ...` present in `dict`.
        fun count(prefix: String, dict: Map<String, Float>): Int {
            var n = 0
            while (dict["$prefix$n"] != null) n += 1
            return n
        }

        /// Return (low-index, high-index, t) for bracketing `x` within `grid`.
        /// Clamps at both ends.
        fun bracket(x: Float, grid: List<Float>): Triple<Int, Int, Float> {
            if (x <= grid.first()) return Triple(0, min(1, grid.size - 1), 0f)
            if (x >= grid.last()) {
                val last = grid.size - 1
                return Triple(max(0, last - 1), last, 1f)
            }
            var lo = 0
            var hi = grid.size - 1
            while (hi - lo > 1) {
                val mid = (hi + lo) / 2
                if (grid[mid] <= x) lo = mid else hi = mid
            }
            val t = (x - grid[lo]) / (grid[hi] - grid[lo])
            return Triple(lo, hi, t)
        }
    }
}
